include/support.h

```diff
@@ -191,4 +191,22 @@ constexpr float coarse_cos(float x)
 	constexpr auto half_pi = static_cast<float>(M_PI_2);
 	return coarse_sin(x + half_pi);
 }
+
+// Use ARRAY_LEN macro to safely calculate number of elements in a C-array.
+// This macro can be used in a constant expressions, even if array is a
+// non-static class member:
+//
+//   constexpr auto n = ARRAY_LEN(my_array);
+//
+template <typename T>
+constexpr size_t static_if_array_then_zero()
+{
+	static_assert(std::is_array<T>::value, "not an array type");
+	return 0;
+}
+
+#define ARRAY_LEN(arr)                                                         \
+	(static_if_array_then_zero<decltype(arr)>() +                          \
+	 (sizeof(arr) / sizeof(arr[0])));
+
 #endif
```

Comment:
This is so nice @dreamer! It should be part of language to have something like this avaiable :-) ie: no runtime over head and safely handles accidents when used on non-arrays.

Reply:
Yeah, I was pretty surprised when I couldnt just use `size()` from `std::array` (because compiler was complaining about `this` pointer and refused to use it in constexpr context - without good reason in my opinion), while standard `sizeof` trick for C tables worked to calculate `constexpr` value … but it's unsafe and might be surprising in some contexts. Combination of C and C++ solution has all the benefits and no downsides :)